Extract samples names

In [1]:
import os

def get_folder_names(directory):
    if not os.path.exists(directory) or not os.path.isdir(directory):
        raise ValueError(f"The path '{directory}' is not a valid directory.")
    
    folder_list = [folder for folder in os.listdir(directory) if os.path.isdir(os.path.join(directory, folder))]
    
    return folder_list

# Example usage
directory_path = "D:\\Nouveau dossier\\data\\data\\MOSEI_RAW_PROCESSED"  # Change this to your target directory
folders = get_folder_names(directory_path)


# Save second-level folder names to a file
with open("D:\\Nouveau dossier\\data\\data\\mosei\\mosei_folder_names.txt", "w") as f:
    for namefolder in folders:
        f.write(namefolder + "\n")


Open the data (meta.pkl)

In [ ]:
import os
import pickle


# Step 1: Load the data from meta.pkl

file_path='D:\\Nouveau dossier\\data\\data\\MOSEI_RAW_PROCESSED\\filtered_data_named_labels.pkl' #replace the path

with open(file_path, 'rb') as f:
    data = pickle.load(f)


distribution of the 6 classes:

In [3]:
import pickle
import pandas as pd


# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize emotion counts
emotion_counts = {emotion: 0 for emotion in emotions}



# Count occurrences of each emotion
for item in data.values():
    for i, label in enumerate(item['label']):
        emotion_counts[emotions[i]] += label

# Return the emotion counts as a dictionary
emotion_counts


{'anger': 4997,
 'disgust': 4097,
 'fear': 1913,
 'happy': 12465,
 'sad': 5998,
 'surprise': 2320}

distribution of only one emotion in samples

In [4]:
import pickle
import pandas as pd

# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize emotion counts
emotion_counts = {emotion: 0 for emotion in emotions}

# Iterate through the data
for item in data.values():
    label = item['label']
    if sum(label) == 1:  # Only one emotion is present
        index = label.index(1)
        emotion_counts[emotions[index]] += 1

# Print the filtered emotion distribution
print(emotion_counts)


{'anger': 987, 'disgust': 400, 'fear': 227, 'happy': 7852, 'sad': 1674, 'surprise': 176}


write them down but for happy only 1700

In [13]:
# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Track how many 'happy' samples have been written
happy_limit = 900
happy_count = 0
sad_limit = 900
sad_count = 0

# List to store selected folder names
selected_folders = []

# Go through each folder
for folder_name, item in data.items():
    label = item['label']
    if sum(label) == 1:  # Ensure it's a single-label sample
        index = label.index(1)
        emotion = emotions[index]

        # Enforce 'happy' limit
        if emotion == 'happy':
            if happy_count >= happy_limit:
                continue
            happy_count += 1
        # Enforce 'sad' limit
        if emotion == 'sad':
            if sad_count >= sad_limit:
                continue
            sad_count += 1    



        selected_folders.append(folder_name)

# Write to text file
with open("D:\\Nouveau dossier\\data\\data\\mosei\\selected_folders.txt", 'w') as f:
    for folder in selected_folders:
        f.write(f"{folder}\n")

print(f"{len(selected_folders)} folder names written to 'selected_folders.txt'")


3590 folder names written to 'selected_folders.txt'


In [8]:
# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize dictionary for counting pairs
pair_counts = {}

# Loop through data
for item in data.values():
    label = item['label']
    if sum(label) == 2:  # Only samples with exactly two emotions
        indices = [i for i, val in enumerate(label) if val == 1]
        e1 = emotions[indices[0]]
        e2 = emotions[indices[1]]
        
        # Sort to treat ('happy', 'sad') same as ('sad', 'happy')
        pair = tuple(sorted([e1, e2]))
        
        if pair in pair_counts:
            pair_counts[pair] += 1
        else:
            pair_counts[pair] = 1

# Print results
for pair, count in pair_counts.items():
    print(f"{pair}: {count}")



('happy', 'sad'): 935
('fear', 'sad'): 322
('anger', 'sad'): 415
('happy', 'surprise'): 776
('anger', 'happy'): 762
('fear', 'happy'): 420
('disgust', 'sad'): 548
('anger', 'disgust'): 796
('anger', 'fear'): 93
('disgust', 'happy'): 294
('anger', 'surprise'): 59
('disgust', 'fear'): 12
('disgust', 'surprise'): 90
('sad', 'surprise'): 115
('fear', 'surprise'): 23


In [ ]:
# Target emotion pairs to save (you can modify this list)


target_pairs = [
   ('fear', 'sad'),
   ('disgust', 'sad'),
   ('anger', 'disgust'),
   ('happy', 'surprise'),
   ('anger', 'fear'),
     ('disgust', 'surprise'),
    ('sad', 'surprise'),
    ('fear', 'surprise'),



]




# Ensure all pairs are sorted alphabetically (to match processing logic)
target_pairs = [tuple(sorted(pair)) for pair in target_pairs]

# Emotion list
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# List to store folder names
selected_folders = []

# Loop through samples
for folder_name, item in data.items():
    label = item['label']
    if sum(label) == 2:
        indices = [i for i, val in enumerate(label) if val == 1]
        pair = tuple(sorted([emotions[indices[0]], emotions[indices[1]]]))
        if pair in target_pairs:
            selected_folders.append(folder_name)

# Write to file
with open("D:\\Nouveau dossier\\data\\data\\mosei\\two_emotion_folders.txt", 'w') as f:
    for folder in selected_folders:
        f.write(f"{folder}\n")

print(f"{len(selected_folders)} folder names written to 'two_emotion_folders.txt'")


2763 folder names written to 'two_emotion_folders.txt'


In [15]:
# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize emotion counter
emotion_counts = {emotion: 0 for emotion in emotions}

# Read folder names from file
with open("D:\\Nouveau dossier\\data\\data\\mosei\\combined.txt", 'r') as f:
    selected_folders = [line.strip() for line in f.readlines()]

# Loop through the selected samples
for folder_name in selected_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        for i, val in enumerate(label):
            if val == 1:
                emotion_counts[emotions[i]] += 1
    else:
        print(f"Warning: {folder_name} not found in data")

# Print the emotion distribution
print("Emotion distribution for selected samples:")
for emotion, count in emotion_counts.items():
    print(f"{emotion}: {count}")


Emotion distribution for selected samples:
anger: 1876
disgust: 1834
fear: 665
happy: 1676
sad: 1885
surprise: 1180


To extract the samples who has only one emotion (sad or happy or anger)

In [16]:
import os

emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Load the keys from a file
def load_keys_from_file(file_path):
    with open(file_path, "r") as file:
        keys = {line.strip() for line in file.readlines()}  # Read keys and remove whitespace
    return keys
#extract the indices of 1 from list
def extract_indices_of_ones(label_vector):
    return [i for i, value in enumerate(label_vector) if value == 1]

# Function to filter keys with exactly 1 or 2 emotions from 'anger', 'happy', 'sad'
def filter_keys(data, keys):
    
    filtered_keys = []

    for key in keys:
        if key in data:
            label_vector = data[key]['label']
            indices = extract_indices_of_ones(label_vector)
         

            # Select keys where exactly 1 or 2 emotions are present
            if indices==[0] or indices==[3] or indices==[4]:
                filtered_keys.append(key)

    return filtered_keys

# Function to compute emotion counts for the filtered keys
def compute_emotion_counts(data, filtered_keys):
    emotion_counts = {emotion: 0 for emotion in emotions}

    for key in filtered_keys:
        for i, label in enumerate(data[key]['label']):
            emotion_counts[emotions[i]] += label

    return emotion_counts

# Example usage
file_path = "D:\\Nouveau dossier\\data\\data\\mosei_folder_names.txt"  # Replace with your file path
keys_from_file = load_keys_from_file(file_path)
filtered_keys = filter_keys(data, keys_from_file)
filtered_emotion_counts = compute_emotion_counts(data, filtered_keys)

# Print results
print("Emotion Counts:", filtered_emotion_counts)

Emotion Counts: {'anger': 987, 'disgust': 0, 'fear': 0, 'happy': 7852, 'sad': 1674, 'surprise': 0}


In [17]:
# Save filtered_keys to a file to get the names of samples who has only one emotion (sad or happy or anger)
with open("D:\\Nouveau dossier\\data\\data\\mosei_folder_names_3emotions.txt", "w") as f:
    for namefolder in filtered_keys:
        f.write(namefolder + "\n")


In [ ]:
#verify the distribution

In [18]:
import pickle


# Load keys from a file
def load_keys_from_file(file_path):
    with open(file_path, "r") as file:
        keys = {line.strip() for line in file.readlines()}  # Read and clean keys
    return keys

# Initialize emotion counts
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']
emotion_counts = {emotion: 0 for emotion in emotions}

# Load keys from file
file_path = "D:\\Nouveau dossier\\data\\data\\mosei_folder_names_3emotions.txt"  # Replace with your actual file path
keys_from_file = load_keys_from_file(file_path)

# Count occurrences of each emotion only for the selected keys
for key in keys_from_file:
    if key in data:  # Ensure the key exists in the dataset
        for i, label in enumerate(data[key]['label']):
            emotion_counts[emotions[i]] += label

# Print the result
print(emotion_counts)


{'anger': 987, 'disgust': 0, 'fear': 0, 'happy': 7852, 'sad': 1674, 'surprise': 0}


In [ ]:
import os
import shutil

# Function to load folder names from file.txt
def load_folder_names(file_path):
    with open(file_path, "r") as file:
        folders = [line.strip() for line in file.readlines() if line.strip()]
    return folders

# Function to move folders to a specified destination directory
def move_folders_to_dataset(source_dir, destination_dir, folder_names):
    if not os.path.exists(destination_dir):  # Ensure the destination directory exists
        os.makedirs(destination_dir)

    for folder_name in folder_names:
        source_folder = os.path.join(source_dir, folder_name)
        destination_folder = os.path.join(destination_dir, folder_name)

        if os.path.exists(source_folder) and os.path.isdir(source_folder):  # Check if folder exists
            shutil.move(source_folder, destination_folder)
            print(f"Moved: {source_folder} → {destination_folder}")
        else:
            print(f"Folder not found: {source_folder}")

# Example Usage
file_path = r"D:\Nouveau dossier\data\data\mosei_folder_names_3emotions.txt"  # Path to file containing folder names
source_directory = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"  # Source folder
destination_directory = r"D:\Nouveau dossier\data\data\Dataset"  # Ensure Dataset is outside source_directory

# Load folder names and move folders
folder_names = load_folder_names(file_path)
move_folders_to_dataset(source_directory, destination_directory, folder_names)

Move the iemocap datasets in the dataset folder with cmu-mosei

In [ ]:
import os
import shutil

# Load valid subfolder names from file.txt
def load_valid_folders(file_path):
    with open(file_path, "r") as file:
        return {line.strip() for line in file.readlines() if line.strip()}  # Read and clean names

# Function to move matching subfolders to Dataset
def move_matching_subfolders(source_root, destination_dir, valid_folders):
    if not os.path.exists(destination_dir):  # Create Dataset folder if not exists
        os.makedirs(destination_dir)

    for session_folder in os.listdir(source_root):  # Iterate over session folders
        session_path = os.path.join(source_root, session_folder)

        if os.path.isdir(session_path):  # Ensure it's a folder
            for subfolder in os.listdir(session_path):  # Iterate over subfolders
                subfolder_path = os.path.join(session_path, subfolder)

                if os.path.isdir(subfolder_path) and subfolder in valid_folders:  # Check match
                    destination_path = os.path.join(destination_dir, subfolder)
                    shutil.move(subfolder_path, destination_path)
                    print(f"Moved: {subfolder_path} → {destination_path}")
                else:
                    print(f"Skipped: {subfolder_path}")

# Example Usage
file_path = r'D:\Nouveau dossier\data\data\IEMOCAP_SPLIT\4emotions_folder_names.txt'  # File with valid subfolder names
source_directory = r"D:\Nouveau dossier\data\data\IEMOCAP_RAW_PROCESSED"  # Root folder
destination_directory = r"D:\Nouveau dossier\data\data\Dataset"  # Target directory

# Load folder names from file.txt
valid_folders = load_valid_folders(file_path)

# Move matching subfolders
move_matching_subfolders(source_directory, destination_directory, valid_folders)

Verify the data we are working on it if iemocap is now in Dataset

In [2]:
import os

# Load folder names from file.txt
def load_folder_names(file_path):
    with open(file_path, "r") as file:
        return {line.strip() for line in file.readlines() if line.strip()}  # Read and clean names

# Verify if folders exist in the Dataset directory
def verify_folders_exist(dataset_dir, folder_names):
    missing_folders = []
    existing_folders = []

    for folder in folder_names:
        folder_path = os.path.join(dataset_dir, folder)
        if os.path.exists(folder_path) and os.path.isdir(folder_path):
            existing_folders.append(folder)
        else:
            missing_folders.append(folder)

    return existing_folders, missing_folders

# Example Usage
file_path = r"D:\Nouveau dossier\data\data\IEMOCAP_SPLIT\4emotions_folder_names.txt"  # File with folder names
dataset_directory = r"D:\Nouveau dossier\data\data\Dataset"  # Directory where folders should exist

# Load folder names from file.txt
folder_names = load_folder_names(file_path)

# Verify folder existence
existing, missing = verify_folders_exist(dataset_directory, folder_names)

# Optional: Print a summary
if not missing:
    print("✅ All folders from file.txt exist in Dataset.")
else:
    print(f"❌ {len(missing)} folders from file.txt are missing in Dataset.")


✅ All folders from file.txt exist in Dataset.


The same thing with cmumosei

Now we want to extract the names of all the folders in Dataset and save them in samples.txt and creat a meta.pkl to save our dataset

Start with iemocap to creat the data.pkl : {key (name of folder): {'text': text, 'label': label, 'dataset': 'iemocap'}}


Paper of confidence based soft feature selection 
Emotion distribution for selected samples:
anger: 1876
disgust: 1834
fear: 665
happy: 1676
sad: 1885
surprise: 1180


In [25]:
import pickle

# Define emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']



# Load the selected folder names from combined.txt
with open("D:\\Nouveau dossier\\data\\data\\mosei\\combined.txt", 'r') as f:
    selected_folders = [line.strip() for line in f.readlines()]

# Create new filtered data dict
filtered_data = {}

for folder_name in selected_folders:
    if folder_name in data:
        text = data[folder_name]['text']
        label_vector = data[folder_name]['label']
        # Convert binary label vector to list of emotion names
        label_names = [emotions[i] for i, val in enumerate(label_vector) if val == 1]
        filtered_data[folder_name] = {
            'text': text,
            'label': data[folder_name]['label']
        }
    else:
        print(f"Warning: {folder_name} not found in original data.")

# Save the new filtered dataset
with open("D:\\Nouveau dossier\\data\\data\\mosei\\filtered_data_named_labels.pkl", "wb") as f:
    pickle.dump(filtered_data, f)

print(f"Filtered data saved with {len(filtered_data)} entries.")


Filtered data saved with 6353 entries.


In [26]:
# Load the original data.pkl
with open("D:\\Nouveau dossier\\data\\data\\mosei\\filtered_data_named_labels.pkl", "rb") as f:
    data = pickle.load(f)
data    

{'-3g5yACwYnA_9': {'text': 'We have many new opportunities through the way things have changed through the years, looking for new niche high-end, value-added products.',
  'label': [0, 0, 1, 0, 1, 0]},
 '-9YyBTjo1zo_3': {'text': "he was so happy to swallow -- he's so happy to eat a golf ball",
  'label': [0, 0, 0, 1, 0, 1]},
 '-9y-fZ3swSY_2': {'text': 'You are hinting that you want her to remove her mustache',
  'label': [0, 0, 0, 1, 0, 1]},
 '-HwX2H8Z4hY_3': {'text': 'Federal spending on education soared eightfold in the past twenty years, rising much faster than inflation',
  'label': [0, 1, 0, 0, 1, 0]},
 '-HwX2H8Z4hY_4': {'text': 'But during the same period, Scholastic Aptitude Test results went down, down and down',
  'label': [0, 1, 0, 0, 1, 0]},
 '-HwX2H8Z4hY_8': {'text': '"Now, that word \'professional\' is becoming more and more of a dirty word for teachers',
  'label': [1, 1, 0, 0, 0, 0]},
 '-HwX2H8Z4hY_9': {'text': "It is a word which the administrator uses when he turns aro

In [3]:
import os
import torch
import numpy as np
import cv2
from PIL import Image
from facenet_pytorch import MTCNN

# Initialize MTCNN with landmark output
device = 'cuda' if torch.cuda.is_available() else 'cpu'
mtcnn = MTCNN(keep_all=False, device=device)

# Paths
combined_txt = r"D:\Nouveau dossier\data\data\mosei\two_emotion_folders2.txt"
data_root = "D:\\Nouveau dossier\\data\\data\\MOSEI_RAW_PROCESSED"

# Parameters
output_size = 128
margin_ratio = 0.3

# Load sample names
with open(combined_txt, 'r') as f:
    sample_names = [line.strip() for line in f.readlines()]

# Track samples with <4 faces
few_faces_samples = []

# Align face using eye landmarks
def align_face(img, landmarks, output_size=128, margin_ratio=0.3):
    # Landmarks: [left_eye, right_eye, nose, mouth_left, mouth_right]
    left_eye, right_eye = landmarks[0], landmarks[1]

    # Compute angle between eyes
    dx = right_eye[0] - left_eye[0]
    dy = right_eye[1] - left_eye[1]
    angle = np.degrees(np.arctan2(dy, dx))

    # Center between eyes
    eye_center = ((left_eye[0] + right_eye[0]) / 2, (left_eye[1] + right_eye[1]) / 2)

    # Convert PIL to OpenCV
    img_cv = np.array(img)
    img_cv = cv2.cvtColor(img_cv, cv2.COLOR_RGB2BGR)

    # Rotate image around eye center
    rot_mat = cv2.getRotationMatrix2D(eye_center, angle, scale=1.0)
    rotated = cv2.warpAffine(img_cv, rot_mat, (img_cv.shape[1], img_cv.shape[0]))

    # Apply same rotation to landmarks
    ones = np.ones((landmarks.shape[0], 1))
    landmarks_hom = np.hstack([landmarks, ones])
    rotated_landmarks = rot_mat @ landmarks_hom.T
    rotated_landmarks = rotated_landmarks.T

    # Bounding box with margin
    x1 = np.min(rotated_landmarks[:, 0])
    y1 = np.min(rotated_landmarks[:, 1])
    x2 = np.max(rotated_landmarks[:, 0])
    y2 = np.max(rotated_landmarks[:, 1])

    cx, cy = (x1 + x2) / 2, (y1 + y2) / 2
    w, h = x2 - x1, y2 - y1
    margin = margin_ratio * max(w, h)

    crop_x1 = int(max(cx - output_size / 2 - margin, 0))
    crop_y1 = int(max(cy - output_size / 2 - margin, 0))
    crop_x2 = int(min(cx + output_size / 2 + margin, rotated.shape[1]))
    crop_y2 = int(min(cy + output_size / 2 + margin, rotated.shape[0]))

    face_crop = rotated[crop_y1:crop_y2, crop_x1:crop_x2]
    face_crop = cv2.resize(face_crop, (output_size, output_size))

    # Convert back to PIL
    face_crop_pil = Image.fromarray(cv2.cvtColor(face_crop, cv2.COLOR_BGR2RGB))
    return face_crop_pil

# Process each sample
for sample in sample_names:
    sample_dir = os.path.join(data_root, sample)
    if not os.path.isdir(sample_dir):
        print(f"[Skip] Not a folder: {sample_dir}")
        continue

    image_files = sorted([f for f in os.listdir(sample_dir) if f.lower().endswith(('.jpg', '.png'))])
    faces_dir = os.path.join(sample_dir, "faces")
    os.makedirs(faces_dir, exist_ok=True)

    face_count = 0

    for img_file in image_files:
        img_path = os.path.join(sample_dir, img_file)
        face_path = os.path.join(faces_dir, img_file)

        try:
            img = Image.open(img_path).convert('RGB')
            boxes, probs, landmarks = mtcnn.detect(img, landmarks=True)

            if boxes is not None and landmarks is not None and probs[0] > 0.90:
                aligned = align_face(img, landmarks[0])  # align first face
                aligned.save(face_path)
                face_count += 1
            else:
                print(f"[No face] {img_file} in {sample}")
        except Exception as e:
            print(f"[Error] {img_file} in {sample}: {e}")

    # Check how many faces were extracted
    if face_count < 4:
        few_faces_samples.append(sample)

# Log samples with few faces
print("\n[Summary] Samples with <4 face frames:")
for s in few_faces_samples:
    print(f"- {s}")

# Optionally save to file
with open("samples_with_few_faces.txt", "w") as f:
    for s in few_faces_samples:
        f.write(s + "\n")


c:\Users\jemma\anaconda3\envs\dataset\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


[No face] image_0.jpg in 4jTUcNlxlYA_0
[No face] image_0.jpg in 6KvG5VbLalY_0
[No face] image_1.jpg in 6KvG5VbLalY_0
[No face] image_21.jpg in A2E78CGi5TU_18
[No face] image_2.jpg in DVvJoMt9o80_2
[No face] image_0.jpg in FWBCTZiijEM_5
[No face] image_1.jpg in FWBCTZiijEM_5
[No face] image_2.jpg in FWBCTZiijEM_5
[No face] image_3.jpg in FWBCTZiijEM_5
[No face] image_4.jpg in FWBCTZiijEM_5
[No face] image_5.jpg in FWBCTZiijEM_5
[No face] image_6.jpg in FWBCTZiijEM_5
[No face] image_7.jpg in FWBCTZiijEM_5
[No face] image_11.jpg in G8aL3bM39co_24
[No face] image_19.jpg in G8aL3bM39co_24
[No face] image_0.jpg in JPT38CA3sGI_0
[No face] image_1.jpg in JPT38CA3sGI_0
[No face] image_0.jpg in JY-krIuEUm4_0
[No face] image_7.jpg in Kn5eKHlPD0k_1
[No face] image_8.jpg in Kn5eKHlPD0k_1
[No face] image_0.jpg in KpMCvUyQCXg_0
[No face] image_0.jpg in NOl0v54DaXo_1
[No face] image_1.jpg in NOl0v54DaXo_1
[No face] image_2.jpg in NOl0v54DaXo_1
[No face] image_3.jpg in NOl0v54DaXo_1
[No face] image_4.j

Remove those that doesn t have faces to recognize

In [2]:
# Define file paths
file1_path = "D:\\Nouveau dossier\\data\\data\\mosei\\combined.txt"
file2_path = "D:\\Nouveau dossier\\data\\data\\mosei\\nofaces.txt"
output_path = "D:\\Nouveau dossier\\data\\data\\mosei\\finalcombined.txt"

# Step 1: Load names to remove from file2.txt
with open(file2_path, 'r') as f2:
    names_to_remove = set(name.strip() for name in f2 if name.strip())

# Step 2: Process file1.txt and remove unwanted names
with open(file1_path, 'r') as f1, open(output_path, 'w') as out:
    for line in f1:
        original_names = line.strip().split()  # assumes space-separated names
        filtered_names = [name for name in original_names if name not in names_to_remove]
        out.write(' '.join(filtered_names) + '\n')

print(f"Filtered lines written to {output_path}")


Filtered lines written to D:\Nouveau dossier\data\data\mosei\finalcombined.txt


GEt the Minimum number of images in 'faces/' and Minimum lenght of 'audio.wav' files

In [70]:
import os
import torchaudio
import numpy as np

# Paths
name_file = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\file.txt"
data_base = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# Target sample rate
target_sr = 8000

# Storage
audio_lengths = []
frame_counts = []

# Read folder names
with open(name_file, 'r') as f:
    folder_names = [line.strip() for line in f if line.strip()]

# Iterate through each sample
for folder in folder_names:
    folder_path = os.path.join(data_base, folder)
    audio_path = os.path.join(folder_path, "audio.wav")
    faces_path = os.path.join(folder_path, "faces")

    # --- Process audio ---
    if os.path.isfile(audio_path):
        try:
            waveform, sr = torchaudio.load(audio_path)
            if sr != target_sr:
                resampler = torchaudio.transforms.Resample(orig_freq=sr, new_freq=target_sr)
                waveform = resampler(waveform)
            audio_lengths.append(waveform.size(1))
        except Exception as e:
            print(f"[Error] Failed to process audio {audio_path}: {e}")
    else:
        print(f"[Warning] Missing audio: {audio_path}")

    # --- Process faces ---
    if os.path.isdir(faces_path):
        image_files = [f for f in os.listdir(faces_path) if f.lower().endswith(('.jpg', '.jpeg', '.png'))]
        frame_counts.append(len(image_files))
    else:
        print(f"[Warning] Missing faces folder: {faces_path}")

# --- Statistics helper ---
def print_stats(label, data):
    if not data:
        print(f"\n❌ No data found for {label}")
        return
    arr = np.array(data)
    print(f"\n📊 {label} Statistics:")
    print(f"   ➤ Min    : {arr.min()}")
    print(f"   ➤ Max    : {arr.max()}")
    print(f"   ➤ Mean   : {arr.mean():.2f}")
    print(f"   ➤ Median : {np.median(arr):.2f}")
    print(f"   ➤ Std    : {arr.std():.2f}")
    print(f"   ➤ 95th % : {np.percentile(arr, 95):.2f}")

# --- Final results ---
print_stats("Resampled Audio Lengths (at 8kHz, in samples)", audio_lengths)
print_stats("Number of Face Images per Sample", frame_counts)





📊 Resampled Audio Lengths (at 8kHz, in samples) Statistics:
   ➤ Min    : 12710
   ➤ Max    : 678796
   ➤ Mean   : 63998.91
   ➤ Median : 54536.00
   ➤ Std    : 40658.35
   ➤ 95th % : 137064.00

📊 Number of Face Images per Sample Statistics:
   ➤ Min    : 4
   ➤ Max    : 166
   ➤ Mean   : 15.73
   ➤ Median : 13.00
   ➤ Std    : 10.14
   ➤ 95th % : 34.00


In [69]:
# Paths to the individual split files
train_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\train_split.txt"
valid_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\valid_split.txt"
test_path  = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\test_split.txt"

# Output file path
combined_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\file.txt"

# Load all lines and remove duplicates using a set
all_names = set()

for path in [train_path, valid_path, test_path]:
    with open(path, 'r') as f:
        lines = [line.strip() for line in f if line.strip()]
        all_names.update(lines)

# Write the combined set to the new file
with open(combined_path, 'w') as f:
    for name in sorted(all_names):  # optional: sort the names
        f.write(name + '\n')

print(f"✅ Combined {len(all_names)} unique names into: {combined_path}")


✅ Combined 4341 unique names into: D:\Nouveau dossier\data\data\MOSEI_SPLIT\file.txt


In [5]:
import os

# Input paths
name_file = r"D:\Nouveau dossier\data\data\mosei\finalcombined.txt"
data_base = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# Validation lists
missing_folders = []
missing_audio = []
missing_faces = []

# Read folder names from file
with open(name_file, 'r') as f:
    folder_names = [line.strip() for line in f if line.strip()]

# Check conditions
for folder in folder_names:
    folder_path = os.path.join(data_base, folder)

    if not os.path.isdir(folder_path):
        missing_folders.append(folder)
        continue  # skip further checks if the main folder doesn't exist

    # Check for audio.wav
    audio_path = os.path.join(folder_path, "audio.wav")
    if not os.path.isfile(audio_path):
        missing_audio.append(folder)

    # Check if 'faces/' exists and has images
    faces_path = os.path.join(folder_path, "faces")
    if not os.path.isdir(faces_path):
        missing_faces.append(folder)
    else:
        image_files = [
            f for f in os.listdir(faces_path)
            if f.lower().endswith(('.jpg', '.jpeg', '.png'))
        ]
        if len(image_files) == 0:
            missing_faces.append(folder)

# Print results
print(f"\n❌ Missing Folders: {len(missing_folders)}")
print(missing_folders)

print(f"\n❌ Missing audio.wav files: {len(missing_audio)}")
print(missing_audio)

print(f"\n❌ Missing or empty faces folders: {len(missing_faces)}")
print(missing_faces)



❌ Missing Folders: 2787
['--qXJuDtHPw_5', '-3g5yACwYnA_4', '-3nNcZdcdvU_0', '-3nNcZdcdvU_2', '-3nNcZdcdvU_5', '-571d8cVauQ_5', '-6rXp3zJ3kc_8', '-6rXp3zJ3kc_9', '-9YyBTjo1zo_1', '-9y-fZ3swSY_0', '-9y-fZ3swSY_1', '-9y-fZ3swSY_7', '-9y-fZ3swSY_8', '-Alixo7euuU_0', '-Alixo7euuU_1', '-Alixo7euuU_13', '-Alixo7euuU_16', '-Alixo7euuU_2', '-Alixo7euuU_7', '-Eqdz5y4pEY_5', '-HeZS2-Prhc_1', '-HeZS2-Prhc_2', '-HeZS2-Prhc_5', '-HvKLjmsO5U_6', '-HwX2H8Z4hY_1', '-HwX2H8Z4hY_7', '-IUUR2yyNbw_1', '-IUUR2yyNbw_3', '-IUUR2yyNbw_6', '-I_e4mIh0yE_1', '-KCahx2qBOI_4', '-LnuDPiuuZw_1', '-LnuDPiuuZw_2', '-LnuDPiuuZw_7', '-MeTTeMJBNc_13', '-NFrJFQijFE_0', '-NFrJFQijFE_1', '-NFrJFQijFE_8', '-RfYyzHpjk4_2', '-RpZEe4w4fY_0', '-RpZEe4w4fY_2', '-RpZEe4w4fY_4', '-RpZEe4w4fY_6', '-RpZEe4w4fY_8', '-SYSVSQnTnA_0', '-SYSVSQnTnA_1', '-SYSVSQnTnA_3', '-THoVjtIkeU_1', '-THoVjtIkeU_11', '-THoVjtIkeU_12', '-THoVjtIkeU_2', '-THoVjtIkeU_6', '-UacrmKiTn4_6', '-UuX1xuaiiE_0', '-UuX1xuaiiE_1', '-UuX1xuaiiE_2', '-UuX1xuaiiE_3', 

In [6]:
import zipfile

# Path to your .zip archive
zip_path = r"D:\Nouveau dossier\data.zip"

# List of folder names that were missing on disk


# Path prefix inside the zip
zip_internal_prefix = "data/data/MOSEI_RAW_PROCESSED/"

# Store found and not-found
found_in_zip = []
not_found_in_zip = []

with zipfile.ZipFile(zip_path, 'r') as zipf:
    # Get all folder paths in the zip
    all_zip_paths = set(zipf.namelist())

    for folder in missing_folders:
        folder_prefix = f"{zip_internal_prefix}{folder}/"
        # Check if any file or subpath starts with this folder path
        found = any(path.startswith(folder_prefix) for path in all_zip_paths)

        if found:
            found_in_zip.append(folder)
        else:
            not_found_in_zip.append(folder)

# Final results
print(f"\n✅ Found in zip: {len(found_in_zip)}")
print(found_in_zip)

print(f"\n❌ Still missing (not in zip): {len(not_found_in_zip)}")
print(not_found_in_zip)



✅ Found in zip: 0
[]

❌ Still missing (not in zip): 2787
['--qXJuDtHPw_5', '-3g5yACwYnA_4', '-3nNcZdcdvU_0', '-3nNcZdcdvU_2', '-3nNcZdcdvU_5', '-571d8cVauQ_5', '-6rXp3zJ3kc_8', '-6rXp3zJ3kc_9', '-9YyBTjo1zo_1', '-9y-fZ3swSY_0', '-9y-fZ3swSY_1', '-9y-fZ3swSY_7', '-9y-fZ3swSY_8', '-Alixo7euuU_0', '-Alixo7euuU_1', '-Alixo7euuU_13', '-Alixo7euuU_16', '-Alixo7euuU_2', '-Alixo7euuU_7', '-Eqdz5y4pEY_5', '-HeZS2-Prhc_1', '-HeZS2-Prhc_2', '-HeZS2-Prhc_5', '-HvKLjmsO5U_6', '-HwX2H8Z4hY_1', '-HwX2H8Z4hY_7', '-IUUR2yyNbw_1', '-IUUR2yyNbw_3', '-IUUR2yyNbw_6', '-I_e4mIh0yE_1', '-KCahx2qBOI_4', '-LnuDPiuuZw_1', '-LnuDPiuuZw_2', '-LnuDPiuuZw_7', '-MeTTeMJBNc_13', '-NFrJFQijFE_0', '-NFrJFQijFE_1', '-NFrJFQijFE_8', '-RfYyzHpjk4_2', '-RpZEe4w4fY_0', '-RpZEe4w4fY_2', '-RpZEe4w4fY_4', '-RpZEe4w4fY_6', '-RpZEe4w4fY_8', '-SYSVSQnTnA_0', '-SYSVSQnTnA_1', '-SYSVSQnTnA_3', '-THoVjtIkeU_1', '-THoVjtIkeU_11', '-THoVjtIkeU_12', '-THoVjtIkeU_2', '-THoVjtIkeU_6', '-UacrmKiTn4_6', '-UuX1xuaiiE_0', '-UuX1xuaiiE_1', '

In [ ]:




# Paths
combined_txt = r"D:\Nouveau dossier\data\data\mosei\finalcombined.txt"


# Load all folder names
with open(combined_txt, 'r') as f:
    all_folders = [line.strip() for line in f if line.strip()]

# Filter out the missing ones
filtered_folders = [folder for folder in all_folders if folder not in missing_folders]

# Overwrite the original file with filtered list
with open(combined_txt, 'w') as f:
    for folder in filtered_folders:
        f.write(folder + '\n')

print(f"✅ Updated {combined_txt} — removed {len(missing_folders)} entries.")


✅ Updated D:\Nouveau dossier\data\data\mosei\finalcombined.txt — removed 2787 entries.


In [6]:
import os
import pickle


# Step 1: Load the data from meta.pkl

file_path='D:\\Nouveau dossier\\data\\data\\mosei\\filtered_data_named_labels.pkl' #replace the path

with open(file_path, 'rb') as f:
    data = pickle.load(f)


# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize emotion counter
emotion_counts = {emotion: 0 for emotion in emotions}

# Read folder names from file
with open("D:\\Nouveau dossier\\data\\data\\mosei\\finalcombined.txt", 'r') as f:
    selected_folders = [line.strip() for line in f.readlines()]

# Loop through the selected samples
for folder_name in selected_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        for i, val in enumerate(label):
            if val == 1:
                emotion_counts[emotions[i]] += 1
    else:
        print(f"Warning: {folder_name} not found in data")

# Print the emotion distribution
print("Emotion distribution for selected samples:")
for emotion, count in emotion_counts.items():
    print(f"{emotion}: {count}")    

Emotion distribution for selected samples:
anger: 860
disgust: 1786
fear: 654
happy: 741
sad: 969
surprise: 1131


In [13]:
import os

# Path to your folder list and base directory
name_file = r"D:\Nouveau dossier\data\data\mosei\finalcombined.txt"
data_base = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# List to store image counts
image_counts = []

# Load folder names
with open(name_file, 'r') as f:
    folder_names = [line.strip() for line in f if line.strip()]

# Count images in faces/ for each folder
for folder in folder_names:
    faces_path = os.path.join(data_base, folder, "faces")

    if os.path.isdir(faces_path):
        image_files = [
            f for f in os.listdir(faces_path)
            if f.lower().endswith(('.jpg', '.jpeg', '.png'))
        ]
        image_counts.append(len(image_files))
    else:
        print(f"[Warning] Missing or invalid faces folder: {faces_path}")

# Final min/max result
if image_counts:
    print("\n📊 Number of images in 'faces/' folders:")
    print(f"✅ Min : {min(image_counts)} images")
    print(f"✅ Max : {max(image_counts)} images")
else:
    print("❌ No valid faces folders found.")



📊 Number of images in 'faces/' folders:
✅ Min : 4 images
✅ Max : 114 images


In [7]:
import os

# Paths
name_file = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\train_split.txt"
data_base = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# List to store image counts
image_counts = []
folders_with_4_images = []

# Read all folder names
with open(name_file, 'r') as f:
    folder_names = [line.strip() for line in f if line.strip()]

# Process each folder
for folder in folder_names:
    faces_path = os.path.join(data_base, folder, "faces")

    if os.path.isdir(faces_path):
        image_files = [
            f for f in os.listdir(faces_path)
            if f.lower().endswith(('.jpg', '.jpeg', '.png'))
        ]
        num_images = len(image_files)
        image_counts.append(num_images)

        if num_images ==0:
            folders_with_4_images.append(folder)
    else:
        print(f"[Warning] Missing faces folder: {faces_path}")

# Results
print("\n📊 Face Image Stats:")
print(f"✅ Min images: {min(image_counts)}")
print(f"✅ Max images: {max(image_counts)}")
print(f"✅ Number of folders with exactly 4 images: {len(folders_with_4_images)}")

# Optional: print or save the list of those folders
#print(folders_with_4_images)




📊 Face Image Stats:
✅ Min images: 4
✅ Max images: 166
✅ Number of folders with exactly 4 images: 0


Split the data


In [14]:
import os
import pickle
import random
from collections import defaultdict

# Paths
combined_path = r"D:\Nouveau dossier\data\data\mosei\finalcombined_filtered.txt"
pkl_path = r"D:\Nouveau dossier\data\data\mosei\filtered_data_named_labels.pkl"
output_dir = r"D:\Nouveau dossier\data\data\mosei"
os.makedirs(output_dir, exist_ok=True)

# Load list of sample folder names
with open(combined_path, 'r') as f:
    folder_names = [line.strip() for line in f if line.strip()]

# Load emotion labels
with open(pkl_path, 'rb') as f:
    label_dict = pickle.load(f)

# Group folders by primary emotion
emotion_to_folders = defaultdict(list)
unlabeled = []

for folder in folder_names:
    if folder in label_dict:
        label_vector = label_dict[folder]['label']
        if 1 in label_vector:
            main_emotion = label_vector.index(1)  # pick first 1 as emotion index
            emotion_to_folders[main_emotion].append(folder)
        else:
            unlabeled.append(folder)
            print(f"[Warning] No active label for {folder}")
    else:
        unlabeled.append(folder)
        print(f"[Warning] Missing from label dict: {folder}")

# Stratified 85/5/10 split per emotion
train_set, val_set, test_set = set(), set(), set()
random.seed(42)

for emotion, folders in emotion_to_folders.items():
    random.shuffle(folders)
    n = len(folders)
    n_train = int(0.85 * n)
    n_val = int(0.05 * n)
    n_test = n - n_train - n_val  # all remaining

    train_set.update(folders[:n_train])
    val_set.update(folders[n_train:n_train + n_val])
    test_set.update(folders[n_train + n_val:])

# Ensure disjoint sets
assert train_set.isdisjoint(val_set)
assert train_set.isdisjoint(test_set)
assert val_set.isdisjoint(test_set)

# Save splits
def save_split(split_set, name):
    path = os.path.join(output_dir, f"{name}_split.txt")
    with open(path, 'w') as f:
        for item in sorted(split_set):
            f.write(item + '\n')
    print(f"✅ Saved {len(split_set)} to {path}")

save_split(train_set, "train")
save_split(val_set, "valid")
save_split(test_set, "test")

# Summary
print(f"\n📊 Processed labeled folders: {len(train_set) + len(val_set) + len(test_set)}")
print(f"❌ Unlabeled or missing: {len(unlabeled)}")



✅ Saved 2530 to D:\Nouveau dossier\data\data\mosei\train_split.txt
✅ Saved 147 to D:\Nouveau dossier\data\data\mosei\valid_split.txt
✅ Saved 303 to D:\Nouveau dossier\data\data\mosei\test_split.txt

📊 Processed labeled folders: 2980
❌ Unlabeled or missing: 0


In [10]:
import os
import pickle


# Step 1: Load the data from meta.pkl

file_path='D:\\Nouveau dossier\\data\\data\\MOSEI_RAW_PROCESSED\\filtered_data_named_labels.pkl' #replace the path

with open(file_path, 'rb') as f:
    data = pickle.load(f)


# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize emotion counter
emotion_counts = {emotion: 0 for emotion in emotions}

# Read folder names from file
with open("D:\\Nouveau dossier\\data\\data\\MOSEI_SPLIT\\test_split.txt", 'r') as f: 
    selected_folders = [line.strip() for line in f.readlines()]

# Loop through the selected samples
for folder_name in selected_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        for i, val in enumerate(label):
            if val == 1:
                emotion_counts[emotions[i]] += 1
    else:
        print(f"Warning: {folder_name} not found in data")

# Print the emotion distribution
print("Emotion distribution for selected samples:")
for emotion, count in emotion_counts.items():
    print(f"{emotion}: {count}")    

Emotion distribution for selected samples:
anger: 86
disgust: 136
fear: 62
happy: 75
sad: 102
surprise: 108


In [12]:
import os
import pickle
import random

# Paths
original_file = r"D:\Nouveau dossier\data\data\mosei\finalcombined.txt"
pkl_file = r"D:\Nouveau dossier\data\data\mosei\filtered_data_named_labels.pkl"
output_file = r"D:\Nouveau dossier\data\data\mosei\finalcombined_filtered.txt"

# Parameters
NUM_PURE_DISGUST_TO_REMOVE = 550
random.seed(42)

# Load folder list
with open(original_file, 'r') as f:
    folder_names = [line.strip() for line in f if line.strip()]

# Load labels
with open(pkl_file, 'rb') as f:
    label_dict = pickle.load(f)

# Containers
pure_disgust = []              # exactly [0, 1, 0, 0, 0, 0]
disgust_and_surprise = []      # where label[1] == 1 and label[5] == 1
final_filtered = []

# Classify folders
for folder in folder_names:
    label_info = label_dict.get(folder)
    if not label_info:
        continue

    label = label_info['label']
    if not isinstance(label, list) or len(label) != 6:
        continue

    if label == [0, 1, 0, 0, 0, 0]:
        pure_disgust.append(folder)
        continue

    if label[1] == 1 and label[5] == 1:
        disgust_and_surprise.append(folder)
        continue

    final_filtered.append(folder)

# Randomly remove 550 from pure disgust
random.shuffle(pure_disgust)
pure_disgust_to_keep = pure_disgust[NUM_PURE_DISGUST_TO_REMOVE:]
final_filtered += pure_disgust_to_keep

# Save final list
with open(output_file, 'w') as f:
    for folder in sorted(final_filtered):
        f.write(folder + '\n')

# Summary
print(f"✅ Final combined file saved: {output_file}")
print(f"🚫 Removed {NUM_PURE_DISGUST_TO_REMOVE} pure disgust samples")
print(f"🚫 Removed {len(disgust_and_surprise)} disgust+surprise samples")
print(f"✅ Total retained samples: {len(final_filtered)}")




✅ Final combined file saved: D:\Nouveau dossier\data\data\mosei\finalcombined_filtered.txt
🚫 Removed 550 pure disgust samples
🚫 Removed 89 disgust+surprise samples
✅ Total retained samples: 2980


Final distribution in finalcombined_filtered.txt and 
Emotion distribution for selected samples:
anger: 731
disgust: 1105
fear: 561
happy: 629
sad: 817
surprise: 883   
this distribution was splited into 85 percent train_split.txt, 5percent valid_split.txt and 10percent test_split.txt

In [4]:
import pickle
import random

# Define target emotion pairs
target_pairs = [
    ('fear', 'happy'),
    ('disgust', 'fear'),
    ('anger', 'sad')
]

target_pairs = [tuple(sorted(pair)) for pair in target_pairs]
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']
anger_sad_limit = 300

# File paths
meta_file = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\meta.pkl"
filtered_file = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"
output_txt = r"D:\Nouveau dossier\data\data\mosei\two_emotion_folders2.txt"

# Load full metadata
with open(meta_file, 'rb') as f:
    meta_data = pickle.load(f)

# Load existing filtered data if exists
try:
    with open(filtered_file, 'rb') as f:
        filtered_data = pickle.load(f)
except FileNotFoundError:
    filtered_data = {}

selected_folders = []
anger_sad_folders = []

# Collect matching folder names
for folder_name, item in meta_data.items():
    label = item['label']
    if sum(label) == 2:
        indices = [i for i, val in enumerate(label) if val == 1]
        pair = tuple(sorted([emotions[i] for i in indices]))
        if pair in target_pairs:
            if pair == ('anger', 'sad'):
                anger_sad_folders.append(folder_name)
            else:
                selected_folders.append(folder_name)

# Limit anger-sad to 300 randomly
if len(anger_sad_folders) > anger_sad_limit:
    anger_sad_folders = random.sample(anger_sad_folders, anger_sad_limit)

# Combine all selected folders
selected_folders.extend(anger_sad_folders)

# Write selected folder names to .txt
with open(output_txt, 'w') as f:
    for folder in selected_folders:
        f.write(f"{folder}\n")

# Add selected items to filtered data without removing old ones
for folder in selected_folders:
    if folder in meta_data:
        filtered_data[folder] = meta_data[folder]

# Save updated filtered data
with open(filtered_file, 'wb') as f:
    pickle.dump(filtered_data, f)

print(f"{len(selected_folders)} folders written to text and appended to 'filtered_data_named_labels.pkl'")



732 folders written to text and appended to 'filtered_data_named_labels.pkl'


In [25]:
# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize dictionary for counting pairs
pair_counts = {}
data=filtered_data
# Loop through data
for item in data.values():
    label = item['label']
    if sum(label) == 2:  # Only samples with exactly two emotions
        indices = [i for i, val in enumerate(label) if val == 1]
        e1 = emotions[indices[0]]
        e2 = emotions[indices[1]]
        
        # Sort to treat ('happy', 'sad') same as ('sad', 'happy')
        pair = tuple(sorted([e1, e2]))
        
        if pair in pair_counts:
            pair_counts[pair] += 1
        else:
            pair_counts[pair] = 1

# Print results
for pair, count in pair_counts.items():
    print(f"{pair}: {count}")


NameError: name 'filtered_data' is not defined

In [10]:
import random

# File paths
new_folders_path = r"D:\Nouveau dossier\data\data\mosei\two_emotion_folders2.txt"
train_split_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\train_split.txt"

# Load new folder names
with open(new_folders_path, 'r') as f:
    new_folders = [line.strip() for line in f if line.strip()]

# Shuffle them to ensure random insertion order
random.shuffle(new_folders)

# Append shuffled names to train_split.txt
with open(train_split_path, 'a') as f:
    for folder in new_folders:
        f.write(f"{folder}\n")

print(f"{len(new_folders)} new folders shuffled and appended to 'train_split.txt'")


732 new folders shuffled and appended to 'train_split.txt'


In [54]:
import os
import pickle


# Step 1: Load the data from meta.pkl

file_path= r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"

with open(file_path, 'rb') as f:
    data = pickle.load(f)


# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize emotion counter
emotion_counts = {emotion: 0 for emotion in emotions}

# Read folder names from file
with open("D:\\Nouveau dossier\\data\\data\\MOSEI_SPLIT\\test_split.txt", 'r') as f: 
    selected_folders = [line.strip() for line in f.readlines()]

# Loop through the selected samples
for folder_name in selected_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        for i, val in enumerate(label):
            if val == 1:
                emotion_counts[emotions[i]] += 1
    else:
        print(f"Warning: {folder_name} not found in data")

# Print the emotion distribution
print("Emotion distribution for selected samples:")
for emotion, count in emotion_counts.items():
    print(f"{emotion}: {count}")    

Emotion distribution for selected samples:
anger: 86
disgust: 136
fear: 62
happy: 75
sad: 102
surprise: 108


In [61]:
import os
import pickle

# Path to your filtered labeled data
file_path = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"

# Load data
with open(file_path, 'rb') as f:
    data = pickle.load(f)

# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize counter for samples with only one emotion
single_emotion_counts = {emotion: 0 for emotion in emotions}
total_one_emotion_samples = 0

# Load selected folder names from train split
with open("D:\\Nouveau dossier\\data\\data\\MOSEI_SPLIT\\train_split.txt", 'r') as f: 
    selected_folders = [line.strip() for line in f.readlines()]

# Loop through samples
for folder_name in selected_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        if sum(label) == 1:  # Only one emotion is active
            idx = label.index(1)
            
            emotion = emotions[idx]
            single_emotion_counts[emotion] += 1
            total_one_emotion_samples += 1
    else:
        print(f"Warning: {folder_name} not found in data")

# Print the emotion distribution for one-emotion samples
print("Samples with exactly one active emotion:")
for emotion, count in single_emotion_counts.items():
    print(f"{emotion}: {count}")

print(f"\nTotal one-emotion samples: {total_one_emotion_samples}")



Samples with exactly one active emotion:
anger: 0
disgust: 0
fear: 195
happy: 200
sad: 200
surprise: 139

Total one-emotion samples: 734


In [60]:
import pickle

# Path to the labeled MOSEI data
file_path = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"

# Load data
with open(file_path, 'rb') as f:
    data = pickle.load(f)

# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Counters
single_emotion_counts = {emotion: 0 for emotion in emotions}
total_one_emotion_samples = 0
multi_emotion_samples = 0

# Iterate over all samples
for folder_name, item in data.items():
    label = item['label']
    if sum(label) == 1:
        idx = label.index(1)
        emotion = emotions[idx]
        single_emotion_counts[emotion] += 1
        total_one_emotion_samples += 1
    elif sum(label) > 1:
        multi_emotion_samples += 1

# Display results
print("Distribution of single-emotion (one-hot) samples:")
for emotion, count in single_emotion_counts.items():
    print(f"{emotion}: {count}")

print(f"\nTotal one-emotion samples: {total_one_emotion_samples}")
print(f"Total multi-emotion samples: {multi_emotion_samples}")
print(f"Total samples in dataset: {len(data)}")



Distribution of single-emotion (one-hot) samples:
anger: 987
disgust: 400
fear: 227
happy: 900
sad: 900
surprise: 176

Total one-emotion samples: 3590
Total multi-emotion samples: 3495
Total samples in dataset: 7085


In [40]:
import pickle

# Load MOSEI labeled data
file_path = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"
with open(file_path, 'rb') as f:
    data = pickle.load(f)

# Get all one-emotion samples
one_emotion_sample_names = [key for key, item in data.items() if sum(item['label']) == 1]

# Load all split files into a set
def load_split(file_path):
    with open(file_path, 'r') as f:
        return set(line.strip() for line in f)

train_split = load_split(r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\train_split.txt")
valid_split = load_split(r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\valid_split.txt")
test_split  = load_split(r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\test_split.txt")

# Combine all split names
all_split_names = train_split.union(valid_split).union(test_split)

# Find one-emotion samples not in any split
not_in_any_split = [name for name in one_emotion_sample_names if name not in all_split_names]

# Save the result
output_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\one_emotion_not_in_any_split.txt"
with open(output_path, 'w') as f:
    for name in not_in_any_split:
        f.write(name + '\n')

# Print summary
print(f"Total one-emotion samples: {len(one_emotion_sample_names)}")
print(f"One-emotion samples NOT in any split: {len(not_in_any_split)}")
print(f"Saved to: {output_path}")




Total one-emotion samples: 3590
One-emotion samples NOT in any split: 3203
Saved to: D:\Nouveau dossier\data\data\MOSEI_SPLIT\one_emotion_not_in_any_split.txt


In [ ]:
import os

# Path to your directory with folders
folder_root = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# Load names from file
one_emotion_file = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\one_emotion_not_in_any_split.txt"
with open(one_emotion_file, 'r') as f:
    one_emotion_names = [line.strip() for line in f.readlines()]

# Get list of actual folders in the directory
actual_folders = set(os.listdir(folder_root))

# Find how many names match folders
found_folders = [name for name in one_emotion_names if name in actual_folders]
not_found_folders = [name for name in one_emotion_names if name not in actual_folders]

# Print results
print(f"Total names in file: {len(one_emotion_names)}")
print(f"Found as folders in directory: {len(found_folders)}")
print(f"Not found as folders: {len(not_found_folders)}")

# Optional: save the not found list
output_missing = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\missing_folders.txt"
with open(output_missing, 'w') as f:
    for name in not_found_folders:
        f.write(name + '\n')

print(f"Missing folder names saved to: {output_missing}")


Total names in file: 3203
Found as folders in directory: 2296
Not found as folders: 907
Missing folder names saved to: D:\Nouveau dossier\data\data\MOSEI_SPLIT\missing_folders.txt


In [45]:
not_found_folders

['--qXJuDtHPw_5',
 '-3g5yACwYnA_4',
 '-3nNcZdcdvU_0',
 '-3nNcZdcdvU_2',
 '-3nNcZdcdvU_5',
 '-571d8cVauQ_5',
 '-6rXp3zJ3kc_8',
 '-6rXp3zJ3kc_9',
 '-9YyBTjo1zo_1',
 '-9y-fZ3swSY_0',
 '-9y-fZ3swSY_1',
 '-9y-fZ3swSY_7',
 '-9y-fZ3swSY_8',
 '-Alixo7euuU_0',
 '-Alixo7euuU_1',
 '-Alixo7euuU_13',
 '-Alixo7euuU_16',
 '-Alixo7euuU_2',
 '-Alixo7euuU_7',
 '-Eqdz5y4pEY_5',
 '-HeZS2-Prhc_1',
 '-HeZS2-Prhc_2',
 '-HeZS2-Prhc_5',
 '-HvKLjmsO5U_6',
 '-HwX2H8Z4hY_1',
 '-HwX2H8Z4hY_7',
 '-IUUR2yyNbw_1',
 '-IUUR2yyNbw_3',
 '-IUUR2yyNbw_6',
 '-I_e4mIh0yE_1',
 '-KCahx2qBOI_4',
 '-LnuDPiuuZw_1',
 '-LnuDPiuuZw_2',
 '-LnuDPiuuZw_7',
 '-MeTTeMJBNc_13',
 '-NFrJFQijFE_0',
 '-NFrJFQijFE_1',
 '-NFrJFQijFE_8',
 '-RfYyzHpjk4_2',
 '-RpZEe4w4fY_0',
 '-RpZEe4w4fY_2',
 '-RpZEe4w4fY_4',
 '-RpZEe4w4fY_6',
 '-RpZEe4w4fY_8',
 '-SYSVSQnTnA_0',
 '-SYSVSQnTnA_1',
 '-SYSVSQnTnA_3',
 '-THoVjtIkeU_1',
 '-THoVjtIkeU_11',
 '-THoVjtIkeU_12',
 '-THoVjtIkeU_2',
 '-THoVjtIkeU_6',
 '-UacrmKiTn4_6',
 '-UuX1xuaiiE_0',
 '-UuX1xuaiiE_1',
 '-Uu

In [43]:
import os

# Base path to folder containing all MOSEI samples
base_path = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# Load the valid folder names from the file
one_emotion_file = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\one_emotion_not_in_any_split.txt"
with open(one_emotion_file, 'r') as f:
    one_emotion_names = [line.strip() for line in f.readlines()]

# Actual folders in the dataset directory
actual_folders = set(os.listdir(base_path))

# Filter: folders in the directory and in the one_emotion list
found_folders = [name for name in one_emotion_names if name in actual_folders]

# Remove the specific folder
if '0SfejBuLgFo_3' in found_folders:
    found_folders.remove('0SfejBuLgFo_3')

# Now filter those with at least 4 images in their 'faces' folder
valid_folders_with_faces = []
for folder_name in found_folders:
    faces_dir = os.path.join(base_path, folder_name, 'faces')
    if os.path.isdir(faces_dir):
        image_files = [f for f in os.listdir(faces_dir) if f.lower().endswith(('.jpg', '.png'))]
        if len(image_files) >= 4:
            valid_folders_with_faces.append(folder_name)

# Save the result to a file
output_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\filtered_faces_min4.txt"
with open(output_path, 'w') as f:
    for name in valid_folders_with_faces:
        f.write(name + '\n')

# Print summary
print(f"Initial matched folders: {len(found_folders)}")
print(f"After filtering '0SfejBuLgFo_3' and min 4 face images: {len(valid_folders_with_faces)}")
print(f"Filtered folder list saved to: {output_path}")


Initial matched folders: 415
After filtering '0SfejBuLgFo_3' and min 4 face images: 390
Filtered folder list saved to: D:\Nouveau dossier\data\data\MOSEI_SPLIT\filtered_faces_min4.txt


In [48]:
import pickle

# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']
emotion_counts = {emotion: 0 for emotion in emotions}

# Load label data
label_path = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"
with open(label_path, 'rb') as f:
    data = pickle.load(f)

# Load filtered folder names (those with >= 4 face images)
valid_folders_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\filtered_faces_min4.txt"
with open(valid_folders_path, 'r') as f:
    valid_folders_with_faces = [line.strip() for line in f.readlines()]

# Count emotion occurrences
for folder_name in found_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        if sum(label) == 1:
            emotion_index = label.index(1)
            emotion = emotions[emotion_index]
            emotion_counts[emotion] += 1
    else:
        print(f"Warning: {folder_name} not found in label data")

# Print the emotion distribution
print("Emotion distribution (one-hot only) for folders with ≥4 face images:")
for emotion, count in emotion_counts.items():
    print(f"{emotion}: {count}")


Emotion distribution (one-hot only) for folders with ≥4 face images:
anger: 987
disgust: 0
fear: 0
happy: 409
sad: 900
surprise: 0


In [47]:
import os
import shutil

# Existing list of folder names you want to copy
# Ensure `found_folders` is already defined in your session
# Example: found_folders = ['abc_1', 'def_2', ...]

# Define source and destination base paths
source_root = r"D:\Nouveau dossier\data\bigdata\Dataset"
destination_root = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# Track how many folders were copied
copied_count = 0

for folder_name in found_folders:
    src = os.path.join(source_root, folder_name)
    dst = os.path.join(destination_root, folder_name)

    if not os.path.exists(src):
        print(f"Source folder not found: {src}")
        continue

    if os.path.exists(dst):
        print(f"Destination already exists, skipping: {dst}")
        continue

    shutil.copytree(src, dst)
    copied_count += 1

print(f"✅ Copied {copied_count} folders from Dataset to MOSEI_RAW_PROCESSED.")


✅ Copied 2296 folders from Dataset to MOSEI_RAW_PROCESSED.


In [49]:
import os
import torch
import numpy as np
import cv2
from PIL import Image
from facenet_pytorch import MTCNN
import shutil

# Device setup
device = 'cuda' if torch.cuda.is_available() else 'cpu'
mtcnn = MTCNN(keep_all=False, device=device)

# Paths
data_root = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED"

# You must already have the list `found_folders`
# Example: found_folders = ['abc_1', 'xyz_3', ...]

# Parameters
output_size = 128
margin_ratio = 0.3

# Track samples with <4 faces
few_faces_samples = []

# Align face using eye landmarks
def align_face(img, landmarks, output_size=128, margin_ratio=0.3):
    left_eye, right_eye = landmarks[0], landmarks[1]
    dx = right_eye[0] - left_eye[0]
    dy = right_eye[1] - left_eye[1]
    angle = np.degrees(np.arctan2(dy, dx))
    eye_center = ((left_eye[0] + right_eye[0]) / 2, (left_eye[1] + right_eye[1]) / 2)
    img_cv = np.array(img)
    img_cv = cv2.cvtColor(img_cv, cv2.COLOR_RGB2BGR)
    rot_mat = cv2.getRotationMatrix2D(eye_center, angle, scale=1.0)
    rotated = cv2.warpAffine(img_cv, rot_mat, (img_cv.shape[1], img_cv.shape[0]))

    ones = np.ones((landmarks.shape[0], 1))
    landmarks_hom = np.hstack([landmarks, ones])
    rotated_landmarks = rot_mat @ landmarks_hom.T
    rotated_landmarks = rotated_landmarks.T

    x1 = np.min(rotated_landmarks[:, 0])
    y1 = np.min(rotated_landmarks[:, 1])
    x2 = np.max(rotated_landmarks[:, 0])
    y2 = np.max(rotated_landmarks[:, 1])
    cx, cy = (x1 + x2) / 2, (y1 + y2) / 2
    w, h = x2 - x1, y2 - y1
    margin = margin_ratio * max(w, h)

    crop_x1 = int(max(cx - output_size / 2 - margin, 0))
    crop_y1 = int(max(cy - output_size / 2 - margin, 0))
    crop_x2 = int(min(cx + output_size / 2 + margin, rotated.shape[1]))
    crop_y2 = int(min(cy + output_size / 2 + margin, rotated.shape[0]))

    face_crop = rotated[crop_y1:crop_y2, crop_x1:crop_x2]
    face_crop = cv2.resize(face_crop, (output_size, output_size))
    face_crop_pil = Image.fromarray(cv2.cvtColor(face_crop, cv2.COLOR_BGR2RGB))
    return face_crop_pil

# Process each sample in found_folders
for sample in found_folders:
    sample_dir = os.path.join(data_root, sample)
    if not os.path.isdir(sample_dir):
        print(f"[Skip] Not a folder: {sample_dir}")
        continue

    # Overwrite 'faces/' folder if it exists
    faces_dir = os.path.join(sample_dir, "faces")
    if os.path.exists(faces_dir):
        shutil.rmtree(faces_dir)
    os.makedirs(faces_dir, exist_ok=True)

    # List image frames
    image_files = sorted([f for f in os.listdir(sample_dir)
                          if f.lower().endswith(('.jpg', '.png')) and os.path.isfile(os.path.join(sample_dir, f))])
    face_count = 0

    for img_file in image_files:
        img_path = os.path.join(sample_dir, img_file)
        face_path = os.path.join(faces_dir, img_file)

        try:
            img = Image.open(img_path).convert('RGB')
            boxes, probs, landmarks = mtcnn.detect(img, landmarks=True)

            if boxes is not None and landmarks is not None and probs[0] > 0.90:
                aligned = align_face(img, landmarks[0])
                aligned.save(face_path)
                face_count += 1
            else:
                print(f"[No face] {img_file} in {sample}")
        except Exception as e:
            print(f"[Error] {img_file} in {sample}: {e}")

    if face_count < 4:
        few_faces_samples.append(sample)

# Log and save the result
print("\n[Summary] Samples with <4 aligned face frames:")
for s in few_faces_samples:
    print(f"- {s}")

with open("samples_with_few_faces.txt", "w") as f:
    for s in few_faces_samples:
        f.write(s + "\n")


[No face] image_0.jpg in -9YyBTjo1zo_1
[No face] image_1.jpg in -9YyBTjo1zo_1
[No face] image_10.jpg in -HwX2H8Z4hY_1
[No face] image_11.jpg in -HwX2H8Z4hY_1
[No face] image_12.jpg in -HwX2H8Z4hY_1
[No face] image_9.jpg in -HwX2H8Z4hY_1
[No face] image_11.jpg in -HwX2H8Z4hY_7
[No face] image_12.jpg in -HwX2H8Z4hY_7
[No face] image_0.jpg in -NFrJFQijFE_0
[No face] image_1.jpg in -NFrJFQijFE_0
[No face] image_2.jpg in -NFrJFQijFE_0
[No face] image_3.jpg in -NFrJFQijFE_0
[No face] image_4.jpg in -NFrJFQijFE_0
[No face] image_5.jpg in -NFrJFQijFE_0
[No face] image_6.jpg in -NFrJFQijFE_0
[No face] image_7.jpg in -NFrJFQijFE_0
[No face] image_0.jpg in -SYSVSQnTnA_0
[No face] image_1.jpg in -SYSVSQnTnA_0
[No face] image_16.jpg in -UuX1xuaiiE_1
[No face] image_18.jpg in -UuX1xuaiiE_1
[No face] image_0.jpg in -UuX1xuaiiE_2
[No face] image_0.jpg in -dxfTGcXJoc_0
[No face] image_1.jpg in -dxfTGcXJoc_0
[No face] image_2.jpg in -dxfTGcXJoc_0
[No face] image_3.jpg in -dxfTGcXJoc_0
[No face] image_4.

In [50]:
# Remove samples with <4 face images from found_folders
found_folders = [name for name in found_folders if name not in few_faces_samples]

# Print result
print(f"Remaining folders after filtering out few-face samples: {len(found_folders)}")


Remaining folders after filtering out few-face samples: 2218


In [51]:
import pickle

# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']
emotion_counts = {emotion: 0 for emotion in emotions}

# Load label data
label_path = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"
with open(label_path, 'rb') as f:
    data = pickle.load(f)

# Load filtered folder names (those with >= 4 face images)
valid_folders_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\filtered_faces_min4.txt"
with open(valid_folders_path, 'r') as f:
    valid_folders_with_faces = [line.strip() for line in f.readlines()]

# Count emotion occurrences
for folder_name in found_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        if sum(label) == 1:
            emotion_index = label.index(1)
            emotion = emotions[emotion_index]
            emotion_counts[emotion] += 1
    else:
        print(f"Warning: {folder_name} not found in label data")

# Print the emotion distribution
print("Emotion distribution (one-hot only) for folders with ≥4 face images:")
for emotion, count in emotion_counts.items():
    print(f"{emotion}: {count}")

Emotion distribution (one-hot only) for folders with ≥4 face images:
anger: 952
disgust: 0
fear: 0
happy: 398
sad: 868
surprise: 0


In [65]:
import os
import pickle
import random

# Set seed for reproducibility
random.seed(42)

# Path to MOSEI label file
label_file = r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"

# Paths to split files
train_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\train_split.txt"
valid_path = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\valid_split.txt"
test_path  = r"D:\Nouveau dossier\data\data\MOSEI_SPLIT\test_split.txt"

# Your existing found_folders list must be in memory
# found_folders = [...]

# Load labels
with open(label_file, 'rb') as f:
    data = pickle.load(f)

# Load existing splits
def load_split(path):
    if not os.path.exists(path):
        return set()
    with open(path, 'r') as f:
        return set(line.strip() for line in f.readlines())

train_set = load_split(train_path)
valid_set = load_split(valid_path)
test_set  = load_split(test_path)
already_used = train_set.union(valid_set).union(test_set)

# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Only target anger
emotion_targets = {
    'train': {'anger': 100},
    'valid': {'anger': 20},
    'test':  {'anger': 30}
}

# Bucket one-emotion samples for anger only
anger_bucket = []
for name in found_folders:
    if name in data and sum(data[name]['label']) == 1:
        label_idx = data[name]['label'].index(1)
        emotion = emotions[label_idx]
        if emotion == 'anger' and name not in already_used:
            anger_bucket.append(name)

# Shuffle anger samples
random.shuffle(anger_bucket)

# Allocate helper
def allocate_split(bucket, count):
    selected = bucket[:count]
    del bucket[:count]
    return selected

# Split allocation
split_names = {'train': [], 'valid': [], 'test': []}
split_names['train'].extend(allocate_split(anger_bucket, emotion_targets['train']['anger']))
split_names['valid'].extend(allocate_split(anger_bucket, emotion_targets['valid']['anger']))
split_names['test'].extend(allocate_split(anger_bucket, emotion_targets['test']['anger']))

# Save with deduplication
def append_to_file(path, names, existing_set):
    with open(path, 'a') as f:
        for name in names:
            if name not in existing_set:
                f.write(name + '\n')
                existing_set.add(name)

append_to_file(train_path, split_names['train'], train_set)
append_to_file(valid_path, split_names['valid'], valid_set)
append_to_file(test_path,  split_names['test'],  test_set)

# Report
print(f"Added to train: {len(split_names['train'])}")
print(f"Added to valid: {len(split_names['valid'])}")
print(f"Added to test : {len(split_names['test'])}")



Added to train: 100
Added to valid: 20
Added to test : 30


In [4]:
import os
import pickle


# Step 1: Load the data from meta.pkl

file_path= r"D:\Nouveau dossier\data\data\MOSEI_RAW_PROCESSED\filtered_data_named_labels.pkl"

with open(file_path, 'rb') as f:
    data = pickle.load(f)


# Emotion categories
emotions = ['anger', 'disgust', 'fear', 'happy', 'sad', 'surprise']

# Initialize emotion counter
emotion_counts = {emotion: 0 for emotion in emotions}

# Read folder names from file
with open("D:\\Nouveau dossier\\data\\data\\MOSEI_SPLIT\\train_split.txt", 'r') as f: 
    selected_folders = [line.strip() for line in f.readlines()]

# Loop through the selected samples
for folder_name in selected_folders:
    if folder_name in data:
        label = data[folder_name]['label']
        for i, val in enumerate(label):
            if val == 1:
                emotion_counts[emotions[i]] += 1
    else:
        print(f"Warning: {folder_name} not found in data")

# Print the emotion distribution
print("Emotion distribution for selected samples:")
for emotion, count in emotion_counts.items():
    print(f"{emotion}: {count}")    

Emotion distribution for selected samples:
anger: 1124
disgust: 1117
fear: 989
happy: 1245
sad: 1310
surprise: 883
